### ISSUE TITLE
some questions about implementing RAG.

### ISSUE THREAD

**Issue by zhaoyongchao2019** [LinWeizheDragon/Retrieval-Augmented-Visual-Question-Answering#9]
I noticed that there is code related to RAG-sequence marginalization in your codebase, but you didn't provide the execution command for the RAG model in your README.md. Based on my understanding, if we want to implement the RAG model using your code, we may replace seq_logprobs with rag_logprobs to compute the NLL Loss, and use this loss as the model's loss function for backpropagation. I'm curious about how the RAG model is implemented in your code, and if my understanding is incorrect, could you explain the specific implementation approach?
![image](https://user-images.githubusercontent.com/50958273/231958490-e5f4224f-fb78-4f28-8b0e-61f2c4e90c3b.png)
One last question.I have noticed that RAG, TRiG, and RAVQA models are quite similar, but they differ in some aspects. TRiG does not involve joint training of the Retriever and Generater models, while RAG and RAVQA do. However, they use different loss functions as they address different types of problems. I am not sure if my understanding is correct, so I would like to ask for your opinions.

**Comment by LinWeizheDragon** [comment]
Hi, I implemented the vanilla RAG (the one released by Facebook) in an early stage within my framework. But it turned out to perform very poorly. As a result, I modified the official implementation of RAG to replicate the vanilla RAG baseline. You can check the official implementation released in Huggingface. I produced the input/target csv files and modified the official implementation to make sure that the baseline is accurate. (of course, as you can see from the paper, the performance is still bad with RAG-sequence).

Re your last question, yes, as you have observed, the three models are similar but quite different in both training and inference.
RAG: uses RAG loss to train the system jointly which leverages implicit signals to train the retriever. The loss is derived from K input sequences and 1 target label. At inference time, it uses thorough decoding (see the original paper) to find the answer.
TRiG: an FiD-like system. It takes in K input sequences and generates 1 label when computing the loss, since all K encoder outputs are concatenated before being passed into the decoder.
RAVQA: uses RAVQA loss to train the system jointly with explicit feedback signals from the answer generator. It takes in K input sequences and trains the system to generate K different labels, each corresponding to an answer that appears in the input sequence. At inference time, the model picks the answer with the highest joint probability/confidence (prob of retrieval times prob of answer generation).

**Comment by zhaoyongchao2019** [comment]
Thank you for your response.I have learned a lot from your code and the paper.

**Comment by zhaoyongchao2019** [comment]
@LinWeizheDragon 
Hello,I want to analyze the current bottleneck of this model, but due to GPU limitations, I am unable to train the RAVQA model that includes T5-large. Could you please provide me with the trained model file located at ../Experiments/OKVQA_RA-VQA_FullCorpus/train/saved_model/epoch_xx.ckpt? Thank you very much.
